### PR TITLE
remove duplicate the logic b/w DataFrame API and SQL planning

### PR DESCRIPTION
**Pull request by jiangzhx**
# Which issue does this PR close?

now the count wildcard rules already move to Analyzer #5671
so remove duplicate the logic in SQL planning.

Closes #.

# Rationale for this change
related issues: https://github.com/apache/arrow-datafusion/issues/5473#issuecomment-1454178274
related PR: https://github.com/apache/arrow-datafusion/pull/5671

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/common/src/dfschema.rs [hunk 1]
```diff
@@ -630,9 +630,9 @@ impl ExprSchema for DFSchema {
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub struct DFField {
     /// Optional qualifier (usually a table or relation name)
-    qualifier: Option<OwnedTableReference>,
+    pub qualifier: Option<OwnedTableReference>,
```
Comment (alamb) [inline datafusion/common/src/dfschema.rs:633]: Can you please explain the rationale for this change?

File: datafusion/core/tests/sql/mod.rs [hunk 2]
```diff
@@ -1161,15 +1161,6 @@ async fn try_execute_to_batches(
 /// Execute query and return results as a Vec of RecordBatches
 async fn execute_to_batches(ctx: &SessionContext, sql: &str) -> Vec<RecordBatch> {
     let df = ctx.sql(sql).await.unwrap();
-
-    // We are not really interested in the direct output of optimized_logical_plan
```
Comment (alamb) [inline datafusion/core/tests/sql/mod.rs:1165]: Why was this removed?

**Review by alamb — COMMENTED** [review]
I am sorry for the delay in review. I will try and find more time to review this carefully tomorrow but initially I am surprised that says it removes duplicated logic adds more code than it removes 🤔

**Comment by jiangzhx** [comment]
> I am sorry for the delay in review. I will try and find more time to review this carefully tomorrow but initially I am surprised that says it removes duplicated logic adds more code than it removes 🤔

A large number of test cases are based on SQL, 
like window,union scenarios are not covered dataframe api. 
also the first version of count_wildcard_rule did not cover it.
When I remove the duplicate logic between the DataFrame API and SQL planning, all the problems are exposed.

**Comment by jiangzhx** [comment]
> I am sorry for the delay in review. I will try and find more time to review this carefully tomorrow but initially I am surprised that says it removes duplicated logic adds more code than it removes 🤔


Because when I started to remove the duplicate logic between the DataFrame API and SQL planning, I found that count_wildcard_rule did not cover all scenarios, such as union, window, etc.
Currently, these test cases are all based on SQL, so they are not discovered when running cargo test.
After I covered more scenarios for count_wildcard_rule, it resulted in more code added than removed.

**Comment by alamb** [comment]
> Because when I started to remove the duplicate logic between the DataFrame API and SQL planning, I found that count_wildcard_rule did not cover all scenarios, such as union, window, etc.

This makes sense -- thank you for the explanation @jiangzhx 

Can you please add DataFrame tests the relevant behavior (mostly so that we don't break it in the future by accident)

**Comment by alamb** [comment]
Marking as draft to signify this PR has feedback and is not waiting for another review at the moment.

**Comment by jiangzhx** [comment]
> > Because when I started to remove the duplicate logic between the DataFrame API and SQL planning, I found that count_wildcard_rule did not cover all scenarios, such as union, window, etc.
> 
> This makes sense -- thank you for the explanation @jiangzhx
> 
> Can you please add DataFrame tests the relevant behavior (mostly so that we don't break it in the future by accident)

i add some testcase in tests/dataframe.rs

- test_count_wildcard_on_sort
- test_count_wildcard_on_where_exist
- test_count_wildcard_on_window
- test_count_wildcard_on_aggregate

**Comment by jiangzhx** [comment]
SQL test case Regression in 22.0.0 with count(*) on window

**Comment by jiangzhx** [comment]
split this pr in two part.

1. #6010 
2. remove duplicate the logic b/w DataFrame API and SQL planning

**Comment by alamb** [comment]
Since this has been open for more than a year, closing it down. Feel free to reopen if/when you keep working on it.
